### PR TITLE
[gcs] Update constructor of gcs client

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

GcsClient accepts only redis before. To make it work without redis, we need to be able to pass gcs address to gcs client as well.

In this PR, we add GCS related into into GcsClientOptions so that we can connect to the gcs directly with gcs address.

This  PR is part of GCS bootstrap. In the following PR, we'll add functionality to set the correct GcsClientOptions based on flags.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/worker.py [hunk 1]
```diff
@@ -1452,11 +1453,8 @@ def connect(node,
             "Invalid worker mode. Expected DRIVER, WORKER or LOCAL.")
 
     redis_address, redis_port = node.redis_address.split(":")
-    gcs_options = ray._raylet.GcsClientOptions(
-        redis_address,
-        int(redis_port),
-        node.redis_password,
-    )
+    gcs_options = ray._raylet.GcsClientOptions.from_redis_address(
+        node.redis_address, node.redis_password, False, False, True)
```
Comment (mwtian) [inline python/ray/worker.py:1457]: Use keyword to specify the booleans, and move the comment here?
```
  // As the synchronous and the asynchronous context of redis client is not used in this
  // gcs client. We would not open connection for it by setting `enable_sync_conn` and
  // `enable_async_conn` as false.
```

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 2]
```diff
@@ -88,42 +88,49 @@ GcsClient::GcsClient(
...
-    RAY_CHECK(GetGcsServerAddressFromRedis(
-        redis_client_->GetPrimaryContext()->sync_context(), &current_gcs_server_address_,
-        RayConfig::instance().gcs_service_connect_retries()))
-        << "Failed to get gcs server address when init gcs client.";
+    RAY_CHECK(::RayConfig::instance().gcs_grpc_based_pubsub())
+        << "If using gcs_address to start client, gRPC based pubsub has to be enabled";
+  }
+
+  // Setup gcs server address fetcher
+  if (get_server_address_func_ == nullptr) {
```
Comment (mwtian) [inline src/ray/gcs/gcs_client/gcs_client.cc:112]: How about this?
```
if (get_server_address_func_ != nullptr) {
  // call it once to get gcs address and port
} else if (!options_.gcs_address_.empty()) {
  // set gcs address and port from options_
} else {
  // resolve gcs address and port from redis_client_
}
```
Then we would be able to delete the redis branch after the migration. Also only the redis branch needs retries.

File: src/ray/gcs/gcs_client/gcs_client.h [hunk 3]
```diff
@@ -44,21 +44,28 @@ class GcsClientOptions {
...
-  // GCS server address
-  std::string server_ip_;
-  int server_port_;
+  // Gcs address
+  std::string gcs_address_;
+  int gcs_port_ = 0;
+
+  // redis server address
+  std::string redis_ip_;
+  int redis_port_;
```
Comment (mwtian) [inline src/ray/gcs/gcs_client/gcs_client.h:68]: Default to 0?

File: src/ray/gcs/pubsub/gcs_pub_sub.h [hunk 4]
```diff
@@ -264,9 +265,14 @@ class GcsSubscriber {
   GcsSubscriber(const std::shared_ptr<RedisClient> &redis_client,
                 const rpc::Address &gcs_address,
                 std::unique_ptr<pubsub::Subscriber> subscriber)
-      : pubsub_(std::make_unique<GcsPubSub>(redis_client)),
-        gcs_address_(gcs_address),
-        subscriber_(std::move(subscriber)) {}
+      : gcs_address_(gcs_address), subscriber_(std::move(subscriber)) {
+    if (redis_client) {
```
Comment (mwtian) [inline src/ray/gcs/pubsub/gcs_pub_sub.h:269]: Why not keep the initialization logic as before, and only check when
```
if (pubsub_ != nullptr) {
  RAY_CHECK(...) << ...;
}
```

**Review by mwtian — COMMENTED** [review]
Overall looks good. Some comments about details.

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.cc:112]: The flow is like this:
1. if get_server_address_func_ is set, use this one directly.
2. if get_server_address_func_ is not set, set it
  1. use gcs address
  2. or read from redis


get_server_address_func_ will be used in other places too.

File: src/ray/gcs/pubsub/gcs_pub_sub.h (hunk 4 above)
Comment (iycheng) [reply src/ray/gcs/pubsub/gcs_pub_sub.h:269]: Once gcs address is setup, I don't want the old path to be used anymore.

btw, what do you mean by `if (pubsub_ != nullptr) {` ? It's in the constructor and it's not initialized right?

File: src/ray/core_worker/lib/java/io_ray_runtime_gcs_GlobalStateAccessor.cc [hunk 5]
```diff
@@ -28,8 +29,12 @@ Java_io_ray_runtime_gcs_GlobalStateAccessor_nativeCreateGlobalStateAccessor(
     JNIEnv *env, jobject o, jstring j_redis_address, jstring j_redis_passowrd) {
   std::string redis_address = JavaStringToNativeString(env, j_redis_address);
   std::string redis_password = JavaStringToNativeString(env, j_redis_passowrd);
-  gcs::GlobalStateAccessor *gcs_accessor =
-      new gcs::GlobalStateAccessor(redis_address, redis_password);
+  std::vector<std::string> results;
```
Comment (mwtian) [inline src/ray/core_worker/lib/java/io_ray_runtime_gcs_GlobalStateAccessor.cc:32]: Maybe parsing redis address to create GcsClientOptions can be a common helper function?

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/gcs/gcs_client/gcs_client.cc:112]: I understand the logic in the PR. I just think that
1. Generating `get_server_address_func_` from GCS address in `GcsClientOptions`, then getting gcs address from the function seems unnecessary.
1. Retrying `get_server_address_func_` is only useful when reading GCS address from Redis.
So I'm hoping we only use `get_server_address_func_` when it is set, which does not need to be called elsewhere. What are the places you see where `get_server_address_func_` will be useful?

File: src/ray/gcs/pubsub/gcs_pub_sub.h (hunk 4 above)
Comment (mwtian) [reply src/ray/gcs/pubsub/gcs_pub_sub.h:269]: Ok, I guess we needed to change the initialization logic to keep pubsub_ const. Leaving it non-const is fine.

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.cc:112]: @mwtian `get_server_address_func_` is passed from the constructor, not always generated here. 

Another thing here is that, it doesn't hurt to use get_server_address_func_ to read the address. If it's redis, read from redis, if it's fixed, return the fixed one. We don't need extra if-else for later.

File: src/ray/core_worker/lib/java/io_ray_runtime_gcs_GlobalStateAccessor.cc (hunk 5 above)
Comment (iycheng) [reply src/ray/core_worker/lib/java/io_ray_runtime_gcs_GlobalStateAccessor.cc:32]: No sure we need that or not. It's only in java code right now. In python, it has a converter there.

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/gcs/gcs_client/gcs_client.cc:112]: I have checked the `get_server_address_func_` passed in via constructor, and it was just sending GCS address values. This is the reason retry was not needed before. My proposal does not update `get_server_address_func_`. I still think it is confusing to retry on a function returning in memory values. But if you insist we can keep the logic.

File: src/ray/core_worker/lib/java/io_ray_runtime_gcs_GlobalStateAccessor.cc (hunk 5 above)
Comment (mwtian) [reply src/ray/core_worker/lib/java/io_ray_runtime_gcs_GlobalStateAccessor.cc:32]: I just think the logic is duplicated in `ProcessHelper::CreateGlobalStateAccessor()` too. But the logic is short so keeping them duplicated is fine.

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_client/gcs_client.cc:112]: @mwtian  I'll keep this variable and also the constructor. It looks like we have logic to get gcs address from raylet [code](https://sourcegraph.com/github.com/ray-project/ray@30d3115c4585df3b5cead882b69c7a80c8f023b6/-/blob/src/ray/core_worker/gcs_server_address_updater.cc?L20)

We can get rid of this when we do clean up.

**Review by mwtian — APPROVED** [review]
Looks good. Waiting for CI to pass.

**Pull request merged**